doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,903 @@
+    provisioned.
+
+  - `TLS:HAVE_EL(HANDSHAKE)`: Handled by the handshake layer
+    and forwarded to the record layer to provision keys.
+
+  - `TLS:HANDSHAKE_COMPLETE`: Should be noted as a flag and notification
+    provided to various components.
+
+We choose to model the CSM's state as follows:
+
+  - The `IDLE`, `ACTIVE`, `TERMINATING.CLOSED`, `TERMINATING.DRAINED` and
+    `TERMINATED` states are modelled explicitly as a state variable. However,
+    the substates of `ACTIVE` are not explicitly modelled.
+
+  - The following flags are modelled:
+    - Retry Requested? (+ Original SCID, DCID if so)
+    - Have Sent Any Packet?
+    - Are we currently doing proactive version negotiation?
+    - Have Successfully Received Any Encrypted Packet?
+    - Handshake Completed?
+    - Handshake Confirmed?
+
+  - The following timers are modelled:
+    - PTO Timeout
+    - Terminating Timeout
+    - Idle Timeout
+
+Implementation Plan
+-------------------
+
+- Phase 1: “Steady state only” model which jumps to the `ACTIVE.OPEN`
+  state with a hardcoded key.
+
+  Test plan: Currently uncertain, to be determined.
+
+- Phase 2: “Dummy handshake” model which uses a one-byte protocol
+  as the handshake layer as a standin for TLS 1.3. e.g. a 0x01 byte “represents”
+  a ClientHello, a 0x02 byte “represents” a ServerHello. Keys are fixed.
+
+  Test plan: If feasible, an existing QUIC implementation will be modified to
+  use this protocol and E2E testing will be performed against it. (This
+  can probably be done quickly but an alternate plan may be required if
+  the effort needed turns out be excessive.)
+
+- Phase 3: Final model with TLS 1.3 handshake layer fully plumbed in.
+
+  Test plan: Testing against real world implementations.
+
+Handling of Network I/O
+-----------------------
+
+Our API objectives make handling network I/O tricky.
+Broadly speaking:
+
+ - We want to support both blocking and non-blocking semantics
+   for application use of the libssl APIs.
+
+ - In the case of non-blocking applications, it must be possible
+   for an application to do its own polling and make its own event
+   loop.
+
+Moreover, traditional use of the libssl API allows an application to pass an
+arbitrary BIO to an SSL object; not only that, separate BIOs can be passed for
+the read and write directions. The nature of this BIO can be arbitrary; it could
+be a socket, or a memory buffer.
+
+Implementation of QUIC requires handling of timer events as well as the
+circumstances where a network socket becomes readable or writable. In many cases
+we need to handle these events simultaneously (e.g. wait until a socket becomes
+readable, or a timeout expires, whichever comes first).
+
+### Use of non-blocking I/O
+
+These requirements make it more or less a requirement that we use non-blocking
+network I/O; we need to be able to (at a minimum) have timeouts on recv() calls,
+and make best effort (non blocking) send() and recv() calls.
```

Comment:
This requires a setsockopt call before every read call. This completely defeats the point of us adding a thread safe stateless API (BIO_sendmmsg/BIO_recvmmsg) to BIO_dgram and rules out us making concurrent calls in the future. I'm also not sure there's any coherent way to use this option with sendmmsg/recvmmsg, since it expresses a timeout and not a deadline.